src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/FindPal.cs

```diff
@@ -0,0 +1,275 @@
+                        string oidValue = ConfirmedOidValue(findPal, findValue, OidGroup.Policy);
+                        findPal.FindByApplicationPolicy(oidValue);
+                        break;
+                    }
+                    case X509FindType.FindByCertificatePolicy:
+                    {
+                        string oidValue = ConfirmedOidValue(findPal, findValue, OidGroup.Policy);
+                        findPal.FindByCertificatePolicy(oidValue);
+                        break;
+                    }
+                    case X509FindType.FindByExtension:
+                    {
+                        string oidValue = ConfirmedOidValue(findPal, findValue, OidGroup.ExtensionOrAttribute);
+                        findPal.FindByExtension(oidValue);
+                        break;
+                    }
+                    case X509FindType.FindByKeyUsage:
+                    {
+                        X509KeyUsageFlags keyUsage = ConfirmedX509KeyUsage(findValue);
+                        findPal.FindByKeyUsage(keyUsage);
+                        break;
+                    }
+                    case X509FindType.FindBySubjectKeyIdentifier:
+                    {
+                        byte[] keyIdentifier = ConfirmedCast<string>(findValue).DecodeHexString();
+                        findPal.FindBySubjectKeyIdentifier(keyIdentifier);
+                        break;
+                    }
+                    default:
+                        throw new CryptographicException(SR.Cryptography_X509_InvalidFindType);
+                }
+            }
+           
+            return results;
+        }
+
+        private static T ConfirmedCast<T>(object findValue)
+        {
+            Debug.Assert(findValue != null);
+
+            if (findValue.GetType() != typeof(T))
+                throw new CryptographicException(SR.Cryptography_X509_InvalidFindValue);
+
+            return (T)findValue;
+        }
+
+        private static string ConfirmedOidValue(IFindPal findPal, object findValue, OidGroup oidGroup)
+        {
+            string maybeOid = ConfirmedCast<string>(findValue);
+
+            if (maybeOid.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_InvalidOidValue);
+            }
+
+            return findPal.NormalizeOid(maybeOid, oidGroup);
+        }
+
+        private static X509KeyUsageFlags ConfirmedX509KeyUsage(object findValue)
+        {
+            if (findValue is X509KeyUsageFlags)
+                return (X509KeyUsageFlags)findValue;
+
+            if (findValue is int)
+                return (X509KeyUsageFlags)(int)findValue;
+
+            if (findValue is uint)
+                return (X509KeyUsageFlags)(uint)findValue;
+
+            string findValueString = findValue as string;
+
+            if (findValueString != null)
+            {
+                if (findValueString.Equals("DigitalSignature", StringComparison.OrdinalIgnoreCase))
```

Comment:
Or a `switch` on `findValueString.Length` would be really fast (though, it's more verbose, so might not be worth it depending on how often this code is called)...

Something like:

```c#
switch (findValueString.Length)
{
    case 16:
        switch (char.ToUpperInvariant(findValueString[1]))
        {
            case 'I':
                return ConfirmedX509KeyUsage(findValueString, "DigitalSignature", X509KeyUsageFlags.DigitalSignature);

            case 'A':
                return ConfirmedX509KeyUsage(findValueString, "DataEncipherment", X509KeyUsageFlags.DataEncipherment);
        }
        break;

    case 15:
        return ConfirmedX509KeyUsage(findValueString, "KeyEncipherment", X509KeyUsageFlags.KeyEncipherment);

    case 14:
        return ConfirmedX509KeyUsage(findValueString, "NonRepudiation", X509KeyUsageFlags.NonRepudiation);

    case 12:
        switch (char.ToUpperInvariant(findValueString[0]))
        {
            case 'K':
                return ConfirmedX509KeyUsage(findValueString, "KeyAgreement", X509KeyUsageFlags.KeyAgreement);

            case 'E':
                return ConfirmedX509KeyUsage(findValueString, "EncipherOnly", X509KeyUsageFlags.EncipherOnly);

            case 'D':
                return ConfirmedX509KeyUsage(findValueString, "DecipherOnly", X509KeyUsageFlags.DecipherOnly);
        }
        break;

    case 11:
        return ConfirmedX509KeyUsage(findValueString, "KeyCertSign", X509KeyUsageFlags.KeyCertSign);

    case 7:
        return ConfirmedX509KeyUsage(findValueString, "CrlSign", X509KeyUsageFlags.CrlSign);
}
```

With a helper like this:

```c#
private static X509KeyUsageFlags ConfirmedX509KeyUsage(string findValueString, string flagString, X509KeyUsageFlags flags)
{
    Debug.Assert(findValueString != null);
    Debug.Assert(flagString != null);

    if (flagString.Equals(findValueString, StringComparison.OrdinalIgnoreCase))
    {
        return flags;
    }

    throw new CryptographicException(SR.Cryptography_X509_InvalidFindValue);
}
```

Perhaps with more `Debug.Asserts` to verify the lengths and passed in `flagString` and `flags` match.